### PR TITLE
Add autogluon.common

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Move highly re-usable code to a new autogluon.common module
- autogluon.common will depend on very little, and autogluon.core will depend on autogluon.common

TODO:

- [x] Ensure CI passes
- [x] Move FeatureGenerator and infer_types to autogluon.common (Requires #1376 be merged to avoid conflicts)
- [x] Remove pytest dependency from autogluon.common
- [x] Move all autogluon.core logic used in autogluon.features out of autogluon.core, make autogluon.features depend on autogluon.common instead.
- [x] Merge after #1397 to simplify addressing of merge conflicts.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1386-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1386/8/index.html

**Comment by szha** [comment]
Job PR-1386-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1386/10/index.html

**Comment by szha** [comment]
Job PR-1386-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1386/12/index.html

**Review by gradientsky — APPROVED** [review]

**Comment by szha** [comment]
Job PR-1386-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1386/13/index.html

File: core/src/autogluon/core/_setup_utils.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
     'pandas': '>=1.0.0,<2.0',
     'scikit-learn': '>=1.0.0,<1.1',
     'scipy': '>=1.5.4,<1.7',
+    'psutil': '>=5.7.3,<5.9',  # TODO: Consider capping to <6.0 instead, capping to 5.9 to avoid possible issues.
```
Comment (ddelange) [inline core/src/autogluon/core/_setup_utils.py:22]: @Innixma could this cap be removed now that 5.9 is released?

**Pull request merged**